test/integration/storages_state_test.lua

```diff
@@ -0,0 +1,170 @@
+local fio = require('fio')
+
+local t = require('luatest')
+
+local helpers = require('test.helper')
+
+local fiber = require("fiber")
+
+local pgroup = t.group('storage_info', {
+    {engine = 'memtx'}
+})
+
+local all_storages_initialized = false
+
+local function wait_storages_init(g)
+    local storages_initialized = false
+    local attempts_left = 5
+    local wait_for_init_timeout = 1
+    while (attempts_left > 0 and not storages_initialized) do
+        local results, err = g.cluster.main_server.net_box:call("crud.storage_info", {})
+        t.assert_equals(err, nil, "Error getting storage status")
+        storages_initialized = true
+        for _,v in pairs(results) do
+            if v.status ~= "running" then
+                storages_initialized = false
+            end
+        end
+        if not storages_initialized then
+            fiber.sleep(wait_for_init_timeout)
+            attempts_left = attempts_left-1
+        end
+    end
+    return storages_initialized
+end
+
+pgroup.before_all(function(g)
+    g.cluster = helpers.Cluster:new({
+        datadir = fio.tempdir(),
+        server_command = helpers.entrypoint('srv_select'),
+        use_vshard = true,
+        replicasets = helpers.get_test_replicasets(),
+        env = {
+            ['ENGINE'] = g.params.engine,
+        },
+    })
+    g.cluster:start()
+
+    --[[ wait for storages to initialize
+    This is a workaround for "peer closed" errors for some connections right after the cluster start.
+    Retry is required to give a small timeout to reconnect.
+    ]]--
+    all_storages_initialized = wait_storages_init(g)
+end)
+
+pgroup.after_all(function(g)
+    helpers.stop_cluster(g.cluster)
+    fio.rmtree(g.cluster.datadir)
+end)
+
+pgroup.test_crud_storage_status_of_stopped_servers = function(g)
+    t.assert_equals(all_storages_initialized, true)
```

Comment:
I see in

```lua
t.assert_equals(all_storages_initialized, true)
```
```lua
t.assert_equals(wait_storages_init(g), true)
```

that you rely on tests order and this is not good. Moreover, it seems like you should extract this thing to `before_each` luatest hook. Moreover, if something went wrong in test asserts, instances won't be restarted, so you should extract these parts to `after_each` or `after_test(test_name, ...)` hooks.

Reply:
There is a two separate comments. I should've add two comments instead of one.

1. First test asserts `t.assert_equals(all_storages_initialized, true)` (because `all_storages_initialized` was already computed in `before_all`), all other tests use `t.assert_equals(wait_storages_init(g), true)` (because status has changed and they need to recalculate it). Relying on tests order is not good (for example, if one uses `--shuffle` the logic may break), so let's use honest computation `t.assert_equals(wait_storages_init(g), true)` before each test. And since we want to use it before each test, let's use `before_each` hook instead of copypaste.

2. Some test cases stop servers or disable roles to arrange non-trivial cluster status and validate handle logic. To restore cluster state to a default one, test cases restart servers/enable roles. But since the code is the part of a test case, if some assert will fail, servers won't be restarted/roles won't enabled. Thus next test case will start with inconsistent cluster and will fail too. Let's handle restarts/reenables in `after_test` hooks: they can be used to clean up specific test setups even if test has failed.